### PR TITLE
Support Poetry as Python toolchain

### PR DESCRIPTION
**Pull request by julienp**
# Description

Implement `Toolchain` interface for Poetry.

This does not yet allow the user to select the toolchain during `pulumi new`, but can be used with a template that has explicit poetry support https://github.com/pulumi/templates/pull/795/files

```
pulumi new https://github.com/pulumi/templates/tree/julienp/poetry-template/aws-python-poetry 
```
Stacked on top of https://github.com/pulumi/pulumi/pull/16311

Fixes https://github.com/pulumi/pulumi/issues/16314
Ref https://github.com/pulumi/pulumi/issues/15937

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [x] I have formatted my code using `gofumpt`

### CODE REVIEW

File: sdk/python/toolchain/pip_test.go [hunk 1]
```diff
@@ -119,43 +118,6 @@ func TestRunningPipInVirtualEnvironment(t *testing.T) {
 	}
 }
 
-//nolint:paralleltest // modifies environment variables
```
Comment (julienp) [inline sdk/python/toolchain/pip_test.go:122]: Moved to toolchain_test to cover both pip & poetry

File: sdk/python/toolchain/poetry.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+	return poetryCmd.Run()
+}
+
+func (p *poetry) ListPackages(ctx context.Context, transitive bool) ([]PythonPackage, error) {
+	args := []string{"list", "-v", "--format", "json"}
+	if !transitive {
+		args = append(args, "--not-required")
+	}
+
+	cmd, err := p.ModuleCommand(ctx, "pip", args...)
```
Comment (tgummerer) [inline sdk/python/toolchain/poetry.go:55]: If we are using the poetry toolchain, are we guaranteed to have `pip` installed?

File: pkg/cmd/pulumi/new.go [hunk 3]
```diff
@@ -321,11 +321,15 @@ func runNew(ctx context.Context, args newArgs) error {
...
 	proj.Template = nil
+	// TODO[pulumi/pulumi/issues/16309]: This should be handled by the language runtime.
 	// Workaround for python, most of our templates don't specify a venv but we want to use one
 	if proj.Runtime.Name() == "python" {
-		// If the template does give virtualenv use it, else default to "venv"
-		if _, has := proj.Runtime.Options()["virtualenv"]; !has {
-			proj.Runtime.SetOption("virtualenv", "venv")
+		// If we are not using poetry, backfill the virtualenv option.
+		tc, hasToolchain := proj.Runtime.Options()["toolchain"]
+		if !hasToolchain || tc != "poetry" {
```
Comment (Frassle) [inline pkg/cmd/pulumi/new.go:329]: Feel like this should check for the toolchains which would support venv, rather than checking for not poetry?

**Review by Frassle — COMMENTED** [review]
I'd feel better about this having an integration test in tests/integration/

File: .github/workflows/ci-run-test.yml [hunk 4]
```diff
@@ -214,6 +214,8 @@ jobs:
       - name: Install Python deps
         run: |
           python -m pip install --upgrade pip requests wheel urllib3 chardet
+      - name: Install poetry
+        run: pipx install poetry
```
Comment (tgummerer) [inline .github/workflows/ci-run-test.yml:218]: Where does `pipx` come from?  Can we assume the actions image contains that, or should we add it to the python dependencies above?

**Review by tgummerer — APPROVED** [review]
I got a couple rather naive questions inline, but otherwise looks good to me!

File: sdk/python/toolchain/poetry.go (hunk 2 above)
Comment (julienp) [reply sdk/python/toolchain/poetry.go:55]: Yes, pip is always installed in virtual environments (and any global python installation). https://peps.python.org/pep-0453/

File: .github/workflows/ci-run-test.yml (hunk 4 above)
Comment (julienp) [reply .github/workflows/ci-run-test.yml:218]: Yeah, runner images always include a preinstalled pipx

https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md
https://github.com/actions/runner-images/blob/main/images/macos/macos-14-arm64-Readme.md

File: pkg/cmd/pulumi/new.go (hunk 3 above)
Comment (julienp) [reply pkg/cmd/pulumi/new.go]: Definitely, updated

**Comment by julienp** [comment]
> I'd feel better about this having an integration test in tests/integration/

I split out part of the work into the PR over here https://github.com/pulumi/pulumi/pull/16317 to keep this smaller. That includes an integration test that creates a poetry managed venv.

**Review by Frassle — APPROVED** [review]

**Pull request merged**